OpenAPI-based CEL type library

jiahuif opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
- rewrite the CEL schema library to use `kube-openapi`, based on that of `staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel`, which uses `structural.Structural`.
- two implementations of schema resolver, one using `client-go` OpenAPI v3 discovery mechanism, the other with compiled-in OpenAPI definitions. Compiled-in resolver can be useful for the API server itself while ongoing development of client side CEL evaluation needs the discovery resolver.
- create an adaptor between CRD Structural and OpenAPI schemas. This is to avoid recursive/deep conversion that takes an O(N) time where N is the depth of nesting in a schema.
- change existing library for CRDs to use the adaptor.
- add integration test for above

#### Special notes for your reviewer:
To limit the size of this PR, it does not include the part where the admission plugin uses the library.
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3488
```

cici37 reviewed (commented):
Note: I guess we don't need the compilation and validation files here(might cause confusion). And the code would be tested in https://github.com/kubernetes/kubernetes/pull/113314. Currently the conversion(between object and ref.Val) failed(possible reason of lacking ref resolver?). Will retest after changes are made. Thank you!

Maybe worth to also add builtin type tests in this PR?

jiahuif commented on staging/src/k8s.io/apiserver/pkg/cel/builtins/compilation.go:
```diff
@@ -0,0 +1,273 @@
+//   - nil Program, nil Error: The provided rule was empty so compilation was not attempted
+//
+// perCallLimit was added for testing purpose only. Callers should always use const PerCallLimit as input.
+func Compile(s *spec.Schema, rules []commoncel.ValidationRule, declType *commoncel.DeclType, perCallLimit uint64) ([]CompilationResult, error) {
```
I cannot see a file named `compiler.go`, please paste the full path.

leilajal commented:
/triage accepted

cici37 commented on staging/src/k8s.io/apiserver/pkg/cel/builtins/openapi.go:
```diff
@@ -0,0 +1,97 @@
+		schema.Items.Schema = resolved
+	}
+	if schema.AdditionalProperties != nil && schema.AdditionalProperties.Schema != nil {
+		resolved, err := c.resolveRefs(schema.Items.Schema)
```
schema.AdditionalProperties.Schema?

cici37 replied:
Maybe also add test to catch this?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/builtins/compilation.go:
```diff
@@ -0,0 +1,273 @@
+/*
```
Porting this file would be valuable for adding validation rules on native types (e.g. `// +validationRule: "<cel expression>")` but not for admission control, since we compile admission control quite differently. Are there plans to support validation rules on native types? If not it might be good to shelve this until then.

jiahuif replied:
All unused files removed.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/builtins/validation.go:
```diff
@@ -0,0 +1,295 @@
+/*
```
This file also only needed if for native type validation rules.

jiahuif replied:
All unused files removed.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/builtins/values.go:
```diff
@@ -0,0 +1,702 @@
+// UnstructuredToVal converts a Kubernetes unstructured data element to a CEL Val.
+// The root schema of custom resource schema is expected contain type meta and object meta schemas.
+// If Embedded resources do not contain type meta and object meta schemas, they will be added automatically.
+func UnstructuredToVal(unstructured interface{}, schema *spec.Schema) ref.Val {
```
Given that we primarily intend to support native types with this, should this instead be something like `ObjectToVal(obj runtime.Object) ref.Val` and we can have the implementation look up the schema and convert the runtime.Object to unstructured using `DefaultUnstructuredConverter`? This would allow us to change the implementation in the future (e.g. use a reflection based approach instead of conversion to unstructured) without changing the function signature.

jiahuif replied:
Current implementation let the caller to specify the schema because eventually the schema needs to be cached. It's better for the caller to maintain the cache to avoid having to provide another set of schema cache invalidation.

jpbetz reviewed (commented):
Also add tests showing how this can be used with a native types?

jiahuif committed:
refit CEL typing library

to use OpenAPI schemas.

jiahuif committed:
implement OpenAPI-based schema resolver.

jiahuif committed:
create OWNERS file for cel libs.

jiahuif committed:
add int. test for CEL type resolution.

jiahuif commented:
/retest

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,152 @@
+		if err != nil {
+			return err
+		}
+	}
```
Handle or document the behavior if any of the other fields containing schemas (`oneOf`, `anyOf`, `not`, `patternProperties`, `dependencies`, `definitions`) are encountered? I don't expect us to ever use most of these, but oneOf might be used in the future for unions.

jiahuif replied:
Unfortunately the library handles only the subset that are observed to be used by `kube-openapi`. For `oneOf`, the only use cases would be `int-or-string` and `quantity`, which is correctly handled.

cbandy replied:
> oneOf might be used in the future for unions.

I've wondered but not tried: can union be described in CEL?
```cel
[ has(self.a), has(self.b) ].exists_one(x, x)
```

jiahuif replied:
`exists_one` can be implemented with `filter` and `size`, something like this
`size([has(self.a), has(self.b)].filter(i, i)) == 1`

jiahuif replied:
Nevermind, we already have `exists_one`, your solution is cleaner

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func resolveType(resp *schemaResponse, gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	for _, s := range resp.Components.Schemas {
```
Can `resp` or `resp.Components` ever be nil?

jiahuif replied:
No because `resp` is unmarshalled from JSON, and `Components` is not a pointer. If the returned structure of OpenAPI v3 endpoint ever changes, worst case would be a no-op loop.

jpbetz reviewed (commented):
Generally looks like a clean solution!  Mostly minor comments on the resolver and test coverage.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,152 @@
+		return schema.Ref.String(), true
+	}
+	for _, allOf := range schema.AllOf {
+		if allOf.Ref.GetURL() != nil {
```
Yes, as long as it is generated by api-server. The reason for we wrapping a ref within an `allOf` is preserving the `description` field. ref: https://github.com/kubernetes/kubernetes/issues/106387#issuecomment-967640388

The library handles only the subset of the OpenAPI standard. We support only features that are used by `kube-openapi`.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,152 @@
+	if schema.Ref.GetURL() != nil {
+		return schema.Ref.String(), true
+	}
+	for _, allOf := range schema.AllOf {
```
Because our only use case of `allOf` is to wrap a ref. See https://github.com/kubernetes/kubernetes/issues/106387#issuecomment-967640388

jiahuif commented on test/integration/apiserver/cel/typeresolution_test.go:
```diff
@@ -0,0 +1,362 @@
+		expectEvalErr       bool
+		expectedResult      any
+	}{
+		{
```
I suggest not. This is an integration-level test aimed to test the resolution behavior specifically. `populateRef` would be better left in the unit tests. Other schema-related features are in tests borrowed from the original OpenAPI cel library.

jpbetz replied:
I'm okay with the test coverage being somewhere else, but I would like to ensure we cover these cases somewhere.

jiahuif replied:
Actually `populateRef` is covered in any sub-level field access. `self.spec` would need `DeploymentSpec` to bu populated. I will add more complicated expressions here.

jiahuif replied:
Added. Please take another look when available.

jpbetz commented:
/lgtm

We need approves. @sttts, @deads2k, @liggitt anyone willing to give this a pass? Most of it is a port of existing code, so the new code is mostly isolated to the `implement OpenAPI-based schema resolver.` commit which is quite modest in size.

jiahuif commented:
/assign @liggitt 
Could you take a look when available? Thank you.

liggitt commented:
/test check-dependency-stats

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,164 @@
+
+var _ SchemaResolver = (*ClientDiscoveryResolver)(nil)
+
+func (r *ClientDiscoveryResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	p, err := r.Discovery.OpenAPIV3().Paths()
+	if err != nil {
+		return nil, err
+	}
+	resourcePath := resourcePathFromGV(gvk.GroupVersion())
+	c, ok := p[resourcePath]
+	if !ok {
+		return nil, fmt.Errorf("cannot resolve group version %q: %w", gvk.GroupVersion(), ErrSchemaNotFound)
+	}
+	b, err := c.Schema(runtime.ContentTypeJSON)
```
how often is this called, and in what code path? are we expecting an implementation of the discovery interface that handles caching, etc?

jiahuif replied:
I expect resolved schemas to be cached. For now the resolution would happen for every type check, and, in worst case, for every evaluation. Performance tuning would happen when the feature graduates, however I do not expect this library to change as much. Instead, the results should be cached by the caller.

Yes we do expect the caller to do the caching and use the cached Discovery interface. The main reason is that the admission controller has knowledge of CRDs being changed (with an informer) and should be easier to invalidate.

liggitt replied:
An API server process that needs schema information for a given GVK in order to do type-aware compilation for admission should not need to make a network call to resolve it, or download an openapi schema, or unmarshal it. 

The API server running admission already has the schema information in-process, it should use that rather than pulling from discovery.

Modeling this as a SchemaResolver interface is good, but I would have expected the first two implementations to be one that pulled from compiled-in openapi info (used by kube-apiserver for all its types, kube-aggregator for APIService, apiextensions-apiserver for CustomResourceDefinition and Scale types) and one that converted from CRD-defined schemas (used by apiextensions-apiserver for custom resources with schemas).

We don't need an API server process to evaluate a schema fetched from the network for a type served by another API server, so let's not put that complexity and performance hit in the admission path.

jiahuif replied:
> The API server running admission already has the schema information in-process, it should use that rather than pulling from discovery.

> We don't need an API server process to evaluate a schema fetched from the network for a type served by another API server, so let's not put that complexity and performance hit in the admission path.

This is resolved by the compiled-in definition resolver; the discovery resolver is however kept for client-side where the compiled-in schemas do not exist.

`one that converted from CRD-defined schemas (used by apiextensions-apiserver for custom resources with schemas).`

CRD to Structural is already there, if any change is needed I would consider it out of scope of this PR.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		return nil, err
+	}
+	resp := new(schemaResponse)
+	err = json.Unmarshal(b, resp)
```
@apelisse, is this going to make full use of the unmarshal improvements in https://github.com/kubernetes/kube-openapi/pull/319 or is it going to double unmarshal because we're calling json.Unmarshal first?

jiahuif replied:
I borrowed `kubectl explain`, which seems to be the only place that utilizes OpenAPI v3 discovery client. https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/kubectl/pkg/explain/v2/explain.go#L78

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/maplist.go:
```diff
@@ -0,0 +1,179 @@
+/*
```
same question about how we'll ensure staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/maplist.go and this file don't drift

jiahuif replied:
ditto

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,164 @@
+var _ SchemaResolver = (*ClientDiscoveryResolver)(nil)
+
+func (r *ClientDiscoveryResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	p, err := r.Discovery.OpenAPIV3().Paths()
```
Since the admission handler will only be called for types served from this server process, I'm a little surprised to see us fetching the openapi schemas to resolve the schema info from the network.

jiahuif replied:
In short, to avoid having to re-implement features that apiserver already has. Main reasons for requesting the schema via the loopback are
- unifying built-ins and CRDs. apiserver can already return schemas by a group name. Otherwise, we would need to check if it is built-in, look up compiled-in schemas if built-in, and look up and convert `openAPIV3Schema` if CRD.
- we get caching and grouping for free. OpenAPI v3 endpoint groups a group (as of GVK) and its dependencies neatly in one document.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,164 @@
+
+var _ SchemaResolver = (*ClientDiscoveryResolver)(nil)
+
+func (r *ClientDiscoveryResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
```
> To limit the size of this PR, it does not include the part where the admission plugin uses the library.

This looks like it is trying to prep for typed evaluation, which surprises me... I thought all the cel evaluation from admission was going to be dynamic, since we didn't know a single schema for all the types a policy would handle.

If I set up a validating admission policy that matches `apiGroups:["*"], resources: ["*"]`, will this try to resolve a schema for every type eventually?

jiahuif replied:
No. Schemas are resolved in an as-needed basis. Type checking happens when the CEL expression compiles, which in turn happens every time an expression evaluates against an object, at least as of 1.26. Type checking is optional and can be turned off in a per-expression basis. There is also plan to do an early type checking if the referred type and param can be determined.

Again this part is still being designed. I will cc you on KEP updates and such.

liggitt replied:
making `*spec.Schema` the unifying layer to treat CRDs and built-in types consistently is a plausible goal even if the details of how validating admission will use this are still unclear

moving the existing CRD CEL validation logic from structural.Schema to *spec.Schema seems like a good first step, to avoid duplication and ensure this version still works with the CRD CEL validation

as a follow-up, as we're thinking about how validating admission would use typed evaluation, I would have questions like these:
* since CEL compilation is relatively expensive, we want to amortize compilation cost over many evaluations. can we still do that if we're compiling type-specific versions of CEL admission validation?
* how many distinct type-aware versions of a single CEL admission rule would we compile (at the limit, imagine a cluster with thousands of CRDs defined which configures an `apiGroups:["*"], apiVersions:["*"], resources: ["*"]` rule... will we compile and cache thousands of instances of the CEL program?
* admission rules can (and probably should be) created or updated in advance of the creation of the CRD they protect (or the update of the CRD to add new fields), yet that guarantees the schema will not be available yet, or will not contain the newly defined fields in the CRD. If we're using this at CRD write time to issue warnings, how do we make that a reasonable user experience?
* etc

jiahuif replied:
> moving the existing CRD CEL validation logic from structural.Schema to *spec.Schema seems like a good first step, to avoid duplication and ensure this version still works with the CRD CEL validation

As a recap, the conversion turns out to be too expensive and thus it uses an adoptor instead.

An update after the KEP finalized ([related PR](https://github.com/kubernetes/enhancements/pull/3779)). The design was simplified compared to what I had originally planned. Many of the problems considered during implementation of this library no longer exist.

> since CEL compilation is relatively expensive, we want to amortize compilation cost over many evaluations. can we still do that if we're compiling type-specific versions of CEL admission validation?
Unfortunately the evaluation remains using dynamic type, so yes, compiled expression can be cached with a lifespan same as the policy object;
> how many distinct type-aware versions of a single CEL admission rule would we compile (at the limit, imagine a cluster with thousands of CRDs defined which configures an apiGroups:["*"], apiVersions:["*"], resources: ["*"] rule... will we compile and cache thousands of instances of the CEL program?
ditto. Only one version, which is untyped.
> admission rules can (and probably should be) created or updated in advance of the creation of the CRD they protect (or the update of the CRD to add new fields), yet that guarantees the schema will not be available yet, or will not contain the newly defined fields in the CRD. If we're using this at CRD write time to issue warnings, how do we make that a reasonable user experience?
the warning comes after (1) the schema can be resolved and (2) a problem occurs. This is done as a best effort because ["Backwards-incompatible CRD changes we can treat as not our problem."](https://github.com/kubernetes/enhancements/pull/3779#discussion_r1099293474)

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go:
```diff
@@ -0,0 +1,164 @@
+	resourcePath := resourcePathFromGV(gvk.GroupVersion())
+	c, ok := p[resourcePath]
+	if !ok {
+		return nil, fmt.Errorf("cannot resolve group version %q: %w", gvk.GroupVersion(), ErrSchemaNotFound)
```
if we can't resolve a schema, is that an error or should we fall back to dynamic (untyped) CEL evaluation?

jiahuif replied:
Yes. That's why it wraps the `ErrSchemaNotFound` error.
The design of how ValidatingAdmissionPolicy uses this library to implement type checking is still undergoing but out of scope of the library itself (this PR). I will cc you for the KEP update.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/extensions.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+func isXIntOrString(schema *spec.Schema) bool {
+	return schema.Format == intstr.IntOrString{}.OpenAPISchemaFormat() || isExtension(schema, extIntOrString)
```
There is no constructing of a `IntOrString`, proof https://godbolt.org/z/Pa7sEsoaG

However, let's not rely on compiler behavior and make it clear for future readers. Changing.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/schemas.go:
```diff
@@ -0,0 +1,264 @@
+/*
```
Eventually the one that uses `structural.Structural` would go away, in one of the two ways:
- convert `structural.Structural` to `*spec.Schema`, which we already have https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/kubeopenapi.go#L24
- extract common interfaces of two types.

jiahuif replied:
In the mean time. Because both the CRD validating feature and ValidatingAdmissionPolicy feature have the same group of people as frequent contributors, reviewers, owners, and approvers. Given that OpenAPI schema implementation do not expect much change, we can keep these in sync manually while reviewing.

liggitt replied:
> Eventually the one that uses structural.Structural would go away, in one of the two ways:
>
> * convert structural.Structural to *spec.Schema, which we already have https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/kubeopenapi.go#L24

I would recommend doing that conversion and unifying on constructing CEL types off of *spec.Schema first, not cloning 3+ large and intricate files and trying to reconverge later.

liggitt replied:
that would also make all the existing CRD CEL validation tests exercise this path, which would build confidence in its correctness

jiahuif replied:
I've removed old implementations for Structural. I left original tests intact to avoid breaking anything. However, the CRD validating code couples too strongly with `Structural`, so I kept anything outside the library unchanged with a conversion.

jiahuif commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/values.go:
```diff
@@ -0,0 +1,702 @@
+/*
```
ditto

jiahuif committed:
make structural type use OpenAPI lib.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/maplist_test.go:
```diff
@@ -0,0 +1,297 @@
+/*
```
Curious what others think, but I think I would prefer keeping this new maplist_test.go file but deleting staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/maplist_test.go. While it is good to sanity check that the translation layer is working, I feel like having essentially the same test suite in duplicate will cause drift.

WDYT @liggitt ?

The translation layer changes look good to me otherwise.

jiahuif replied:
Let's keep the tests right now to probe this refactor does not break existing features. We can always remove the tests in another PR.

jpbetz commented:
/lgtm
I lost track of the discussions around the high level schema resolution approach (discovery client vs. internal).  So long as those are resolved, I'm comfortable with this merging. My main focus was the review of the ported CEL code, which all looks good. The code for the discovery resolver all looks functionally correct to me. Test coverage looks good on everything.

jiahuif commented:
Rebased, and addressed some minor formatting issue. [Diff](https://github.com/kubernetes/kubernetes/compare/4791ce8081d7c496fbce31cfca7c15908d2500f0..ef21a3b36406a04a2492a03e3480c43f1b1c3c37)

jiahuif commented:
@liggitt Could you take another look? Thank you.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver/definitions.go:
```diff
@@ -0,0 +1,141 @@
+		return nil, err
+	}
+	result := new(spec.Schema)
+	err = json.Unmarshal(b, result)
```
I'm not quite sure why we're making these json-driven copies, but using the standard json unmarshaler will turn any `int` default values into float64 values ... should we be using the sigs.k8s.io/json decoder?

we've already fixed issues around this in CRD handling (like https://github.com/kubernetes/kubernetes/pull/89833)

wanted to make sure we don't get type issues again (like https://github.com/kubernetes/kubernetes/pull/106051#discussion_r748812421)

jiahuif replied:
Good point, I am working on to either (1) avoid the deep copy or (2) converter, like `ToKubeOpenAPI()`.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver/definitions.go:
```diff
@@ -0,0 +1,141 @@
+	if !ok {
+		return nil, fmt.Errorf("cannot resolve %v: %w", gvk, ErrSchemaNotFound)
+	}
+	result, err := deepCopy(s)
+	if err != nil {
+		return nil, fmt.Errorf("cannot deep copy schema for %v: %v", gvk, err)
+	}
+	err = populateRefs(func(ref string) (*spec.Schema, bool) {
+		// find the schema by the ref string, and return a deep copy
+		def, ok := d.defs[ref]
+		if !ok {
+			return nil, false
+		}
+		s, err := deepCopy(&def.Schema)
```
can you add information about why these are being deep copied (possibly recursively? it's hard to tell how many times the populateRefs function is going to make copies of the same objects)? this seems likely to be expensive

jiahuif replied:
Now the ref resolution will not do a deep copy on the schema.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/values.go:
```diff
@@ -0,0 +1,702 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
(since git doesn't do a great job of showing the file moved and was largely unchanged)

```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package cel
+package openapi
 
 import (
 	"fmt"
@@ -26,24 +26,23 @@ import (
 	"github.com/google/cel-go/common/types/ref"
 	"github.com/google/cel-go/common/types/traits"
 
-	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
-	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/model"
 	"k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apiserver/pkg/cel"
+	"k8s.io/kube-openapi/pkg/validation/spec"
 	"k8s.io/kube-openapi/pkg/validation/strfmt"
 )
 
 // UnstructuredToVal converts a Kubernetes unstructured data element to a CEL Val.
 // The root schema of custom resource schema is expected contain type meta and object meta schemas.
 // If Embedded resources do not contain type meta and object meta schemas, they will be added automatically.
-func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Structural) ref.Val {
+func UnstructuredToVal(unstructured interface{}, schema *spec.Schema) ref.Val {
 	if unstructured == nil {
 		if schema.Nullable {
 			return types.NullValue
 		}
 		return types.NewErr("invalid data, got null for schema with nullable=false")
 	}
-	if schema.XIntOrString {
+	if isXIntOrString(schema) {
 		switch v := unstructured.(type) {
 		case string:
 			return types.String(v)
@@ -56,19 +55,19 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 		}
 		return types.NewErr("invalid data, expected XIntOrString value to be either a string or integer")
 	}
-	if schema.Type == "object" {
+	if schema.Type.Contains("object") {
 		m, ok := unstructured.(map[string]interface{})
 		if !ok {
 			return types.NewErr("invalid data, expected a map for the provided schema with type=object")
 		}
-		if schema.XEmbeddedResource || schema.Properties != nil {
-			if schema.XEmbeddedResource {
-				schema = model.WithTypeAndObjectMeta(schema)
+		if isXEmbeddedResource(schema) || schema.Properties != nil {
+			if isXEmbeddedResource(schema) {
+				schema = WithTypeAndObjectMeta(schema)
 			}
 			return &unstructuredMap{
 				value:  m,
 				schema: schema,
-				propSchema: func(key string) (*structuralschema.Structural, bool) {
+				propSchema: func(key string) (*spec.Schema, bool) {
 					if schema, ok := schema.Properties[key]; ok {
 						return &schema, true
 					}
@@ -76,22 +75,22 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 				},
 			}
 		}
-		if schema.AdditionalProperties != nil && schema.AdditionalProperties.Structural != nil {
+		if schema.AdditionalProperties != nil && schema.AdditionalProperties.Schema != nil {
 			return &unstructuredMap{
 				value:  m,
 				schema: schema,
-				propSchema: func(key string) (*structuralschema.Structural, bool) {
-					return schema.AdditionalProperties.Structural, true
+				propSchema: func(key string) (*spec.Schema, bool) {
+					return schema.AdditionalProperties.Schema, true
 				},
 			}
 		}
 		// A object with x-kubernetes-preserve-unknown-fields but no properties or additionalProperties is treated
 		// as an empty object.
-		if schema.XPreserveUnknownFields {
+		if isXPreserveUnknownFields(schema) {
 			return &unstructuredMap{
 				value:  m,
 				schema: schema,
-				propSchema: func(key string) (*structuralschema.Structural, bool) {
+				propSchema: func(key string) (*spec.Schema, bool) {
 					return nil, false
 				},
 			}
@@ -99,7 +98,7 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 		return types.NewErr("invalid object type, expected either Properties or AdditionalProperties with Allows=true and non-empty Schema")
 	}
 
-	if schema.Type == "array" {
+	if schema.Type.Contains("array") {
 		l, ok := unstructured.([]interface{})
 		if !ok {
 			return types.NewErr("invalid data, expected an array for the provided schema with type=array")
@@ -107,31 +106,30 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 		if schema.Items == nil {
 			return types.NewErr("invalid array type, expected Items with a non-empty Schema")
 		}
-		typedList := unstructuredList{elements: l, itemsSchema: schema.Items}
-		listType := schema.XListType
-		if listType != nil {
-			switch *listType {
+		typedList := unstructuredList{elements: l, itemsSchema: schema.Items.Schema}
+		listType := getXListType(schema)
+		if listType != "" {
+			switch listType {
 			case "map":
-				mapKeys := schema.Extensions.XListMapKeys
+				mapKeys := getXListMapKeys(schema)
 				return &unstructuredMapList{unstructuredList: typedList, escapedKeyProps: escapeKeyProps(mapKeys)}
 			case "set":
 				return &unstructuredSetList{unstructuredList: typedList}
 			case "atomic":
 				return &typedList
 			default:
-				return types.NewErr("invalid x-kubernetes-list-type, expected 'map', 'set' or 'atomic' but got %s", *listType)
+				return types.NewErr("invalid x-kubernetes-list-type, expected 'map', 'set' or 'atomic' but got %s", listType)
 			}
 		}
 		return &typedList
 	}
 
-	if schema.Type == "string" {
+	if schema.Type.Contains("string") {
 		str, ok := unstructured.(string)
 		if !ok {
 			return types.NewErr("invalid data, expected string, got %T", unstructured)
 		}
-		if schema.ValueValidation != nil {
-			switch schema.ValueValidation.Format {
+		switch schema.Format {
 		case "duration":
 			d, err := strfmt.ParseDuration(str)
 			if err != nil {
@@ -158,10 +156,10 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 			}
 			return types.Bytes(base64)
 		}
-		}
+
 		return types.String(str)
 	}
-	if schema.Type == "number" {
+	if schema.Type.Contains("number") {
 		switch v := unstructured.(type) {
 		// float representations of whole numbers (e.g. 1.0, 0.0) can convert to int representations (e.g. 1, 0) in yaml
 		// to json translation, and then get parsed as int64s
@@ -180,7 +178,7 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 			return types.NewErr("invalid data, expected float, got %T", unstructured)
 		}
 	}
-	if schema.Type == "integer" {
+	if schema.Type.Contains("integer") {
 		switch v := unstructured.(type) {
 		case int:
 			return types.Int(v)
@@ -192,7 +190,7 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 			return types.NewErr("invalid data, expected int, got %T", unstructured)
 		}
 	}
-	if schema.Type == "boolean" {
+	if schema.Type.Contains("boolean") {
 		b, ok := unstructured.(bool)
 		if !ok {
 			return types.NewErr("invalid data, expected bool, got %T", unstructured)
@@ -200,7 +198,7 @@ func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Struct
 		return types.Bool(b)
 	}
 
-	if schema.XPreserveUnknownFields {
+	if isXPreserveUnknownFields(schema) {
 		return &unknownPreserved{u: unstructured}
 	}
 
@@ -420,7 +418,7 @@ func (t *unstructuredSetList) Add(other ref.Val) ref.Val {
 // unstructuredList represents an unstructured data instance of an OpenAPI array with x-kubernetes-list-type=atomic (the default).
 type unstructuredList struct {
 	elements    []interface{}
-	itemsSchema *structuralschema.Structural
+	itemsSchema *spec.Schema
 }
 
 var _ = traits.Lister(&unstructuredList{})
@@ -550,9 +548,9 @@ func (t *unstructuredList) Size() ref.Val {
 // unstructuredMap represented an unstructured data instance of an OpenAPI object.
 type unstructuredMap struct {
 	value  map[string]interface{}
-	schema *structuralschema.Structural
+	schema *spec.Schema
 	// propSchema finds the schema to use for a particular map key.
-	propSchema func(key string) (*structuralschema.Structural, bool)
+	propSchema func(key string) (*spec.Schema, bool)
 }
 
 var _ = traits.Mapper(&unstructuredMap{})

```

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/maplist.go:
```diff
@@ -0,0 +1,179 @@
+// keyedItems. If the provided schema is _not_ an array with x-kubernetes-list-type=map, returns an
+// empty mapList.
+func MakeMapList(sts *spec.Schema, items []interface{}) (rv MapList) {
+	if !sts.Type.Contains("array") || getXListType(sts) != "map" || len(getXListMapKeys(sts)) == 0 {
```
this lost the `|| len(items) == 0` check from the original... was that intentional?

jiahuif replied:
It is not. Fixed. Thank you for catching it.
The tests still pass because an empty map list behaves the same as a list map with schema but no items.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/extensions.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func getXListMapKeys(schema *spec.Schema) []string {
+	items, ok := schema.Extensions[extListMapKeys]
```
should `schema.Extensions.GetStringSlice(extListMapKeys)` work for this? it looks like that doesn't handle the `[]string` case, but it could...

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/extensions.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func isXIntOrString(schema *spec.Schema) bool {
+	return schema.Format == intOrStringFormat || isExtension(schema, extIntOrString)
```
curious where `isExtension(schema, extIntOrString)` wasn't sufficient / authoritative

jiahuif replied:
Because built-ins use `Format` field, but CRDs have the extension. This is the first time CEL library encounters this difference. I will add comments.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/values.go:
```diff
@@ -0,0 +1,702 @@
+		}
+		return types.NewErr("invalid data, expected XIntOrString value to be either a string or integer")
+	}
+	if schema.Type.Contains("object") {
```
for all the `schema.Type.Contains("...")` calls, should we be more strict and require schema.Type to exactly equal `[]string{"..."}`? I don't think we support multi-type objects anywhere (question applies globally)

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/schemas.go:
```diff
@@ -0,0 +1,264 @@
+		s = WithTypeAndObjectMeta(s)
+	}
+
+	switch strings.Join(s.Type, ",") {
```
avoid joining, this should be of length 1 and we should be able to switch on `s.Type[0]`

jiahuif replied:
Good point. Added check and use `s.Type[0]` instead.

liggitt commented on staging/src/k8s.io/apiserver/pkg/cel/openapi/schemas.go:
```diff
@@ -0,0 +1,264 @@
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+func SchemaBy(f func(s *spec.Schema)) *spec.Schema {
+	s := new(spec.Schema)
+	f(s)
+	return s
+}
```
this doesn't look used anywhere, unexport it?

actually, I'm not sure what this is getting us... it looks like all the uses pass in a closure... so instead of this:
```go
&spec.Schema{
  Type: []string{"object"},
  AdditionalProperties: &spec.SchemaOrBool{
    Schema: &spec.Schema{
      Type: []string{"string"}
    },
  },
},
```
we can write this:
```go
by(func(s *spec.Schema) {
  s.Type = []string{"object"}
  s.AdditionalProperties = &spec.SchemaOrBool{
    Schema: by(func(s *spec.Schema) {
      s.Type = []string{"string"}
    }),
  }
}),
```

those look... exactly the same length, and the straightforward struct declaration is easier to understand

should we drop the indirection to simplify the code?

jiahuif replied:
The former is actually this 
```go
&spec.Schema{
SchemaProps: spec.SchemaProps{
	Type: []string{"string"},
	AdditionalProperties: &spec.SchemaOrBool{
		Schema: &spec.Schema{
			SchemaProps: spec.SchemaProps{
				Type: []string{"string"},
			},
		},
	},
},
```
Unfortunately, nested object can be directly accessed but not directly constructed.

liggitt replied:
hmm... anything with AdditionalProperties should be an object, right?

looks like there are construction helpers in `spec` so that `spec.MapProperty(spec.StringProperty())` is equivalent to this (and sets Allows:true correctly in SchemaOrBool)

```go
&spec.Schema{
	SchemaProps: spec.SchemaProps{
		Type: []string{"object"},
		AdditionalProperties: &spec.SchemaOrBool{
			Allows: true,
			Schema: &spec.Schema{
				SchemaProps: spec.SchemaProps{
					Type: []string{"string"},
				},
			},
		},
	},
}
```

jiahuif replied:
The fluent helpers do not cover every fields we require, especially extensions. A mixture of construct and helpers is no better than pure construct.

liggitt replied:
fair enough, let's just not export it past this package to avoid picking up other callers

jiahuif replied:
This function saves only one layer of the human readers' reading time, I will go ahead to use the helpers when possible and direct construct otherwise.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/maplist.go:
```diff
@@ -17,162 +17,13 @@ limitations under the License.
-		unkeyedItems: items,
-	}
+func makeMapList(sts *schema.Structural, items []interface{}) (rv celopenapi.MapList) {
+	return celopenapi.MakeMapList(sts.ToKubeOpenAPI(), items)
```
`sts.ToKubeOpenAPI()` is relatively expensive, right? (it looks like it recursively converts the whole subschema to openapi). There are two issues with that:

1. it looks like this is called on every validate of every custom resource and never cached... @jpbetz, I was expecting any schema/structural conversions to be done once per validator instance (either lazily or at validator construction time), not per item validation
2. do we re-call this multiple times at multiple levels? if so, are we reintroducing a performance problem with deeply nested schemas recursing the same trees multiple times?

jiahuif replied:
The runtime cost of `ToKubeOpenAPI()` depends only on the size of the input CRD. There is no "subschema" because CRD cannot embed other types. Preserve Unknown Fields does not increase the depth of the schema either.

1. The validator is indeed cached. It is stored as part of the storage strategy.
https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L52

2. Yes deeply nested schemas are a problem. The worst case scenario would be 
```yaml
    openAPIV3Schema:
      type: object
      properties:
        spec:
          type: object
          x-kubernetes-validations: # ...
          properties:
            nested1:
              type: object
	      x-kubernetes-validations: # ...
	      properties:
	        nested2:
	        type: object
	        x-kubernetes-validations: # ...
	        # keep nesting
``` 
Because of the conversion each time, the time complexity degrades into O(N ^ 2) from O(N)

liggitt replied:
> The validator is indeed cached. It is stored as part of the storage strategy.

The validator is cached, but this wasn't previously being cached along with the validator, right?

jiahuif replied:
Yes you are correct. This part was not cached. It would be costly for the conversion to happen every time. However, the saving grace is that the schema for CRD validator comes with the CR being validated. I am working on a conversion-free solution for the "common ground" between CRD Validation and ValidatingAdmissionPolicy features.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/model/schemas.go:
```diff
@@ -40,152 +33,7 @@ const maxRequestSizeBytes = apiservercel.DefaultMaxRequestSizeBytes
-		return 0
-	}
-	return v
+	return celopenapi.SchemaDeclType(s.ToKubeOpenAPI(), isResourceRoot)
```
same question about whether ToKubeOpenAPI() is called once per validator instance or once per validation

jiahuif replied:
it is called once per validation, every time the variables like `self`, `oldSelf` is constructed.
I've updated the solution to avoid the conversion with the adapter.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go:
```diff
@@ -17,688 +17,16 @@ limitations under the License.
-	}
-
-	return nil, false
+	return celopenapi.UnstructuredToVal(unstructured, schema.ToKubeOpenAPI())
```
same question about whether ToKubeOpenAPI() is called once per validator instance or once per validation

jiahuif replied:
The conversion is per validator, which can be expensive especially for a deeply nested schema. I've changed them to use a adaptor to avoid the conversion.

liggitt commented:
Thanks for the update, keeping the logic in a single place looks a lot more maintainable. The big questions I have are around the json deep copies, and making the structural → schema conversions happen once per validator instance instead of once per validation

jiahuif committed:
adaptor between structral and OpenAPI.

jiahuif committed:
generated: ./hack/update-vendor.sh

jiahuif committed:
avoid deepcopying schemas.

jiahuif commented:
/retest

jiahuif commented:
/retest

jiahuif commented:
/retest

jiahuif commented:
CI is having some issue.
/retest

jiahuif commented:
/test pull-kubernetes-unit

wenjiaswe commented:
/test pull-kubernetes-unit

jpbetz reviewed (commented):
Functionality looks good. The copy-if-changed looks right. The wrapper is nice. Just a few minor comments.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/model/adaptor.go:
```diff
@@ -0,0 +1,149 @@
+	StructuralOrBool *schema.StructuralOrBool
+}
+
+func (sb *StructuralOrBool) Schema() common.Schema {
+	return &Structural{Structural: sb.StructuralOrBool.Structural}
```
I'm not 100% sure this is needed, but it probably doesn't hurt.

jiahuif commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/model/adaptor.go:
```diff
@@ -0,0 +1,149 @@
+	if s.Structural.Properties == nil {
+		return nil
+	}
+	res := make(map[string]common.Schema)
```
Good point! Let me change and doc it.

jpbetz commented:
/lgtm

jiahuif commented:
All comments addressed, could you take another look when available? @liggitt

liggitt commented:
> All comments addressed, could you take another look when available? @liggitt

ack, should be able to look later today

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/maplist.go:
```diff
@@ -17,162 +17,14 @@ limitations under the License.
-		unkeyedItems: items,
-	}
+func makeMapList(sts *schema.Structural, items []interface{}) (rv common.MapList) {
+	return common.MakeMapList(&model.Structural{Structural: sts}, items)
```
same comment here about a follow-up benchmark that calls validate with a CRD and CR that ends up exercising this call so we have visibility to allocations

liggitt reviewed (commented):
/approve

Thanks for all the iterations on this. There are still a couple spots that would be good to be sure we have benchmarks exercising (we might already? I'm not sure), but I'm very happy to see the full deep copies and recursive conversions go away
